File: go/medium/823_binary_trees_with_factors.go
package medium

import "sort"

func numFactoredBinaryTrees(arr []int) int {
	const MOD = 1000000007

	sort.Ints(arr)
	mem := make(map[int]int)
	total := 0

	for i, z := range arr {
		val := 1
		for _, y := range arr[:i] {
			x, r := z/y, z%y

			if r > 0 {
				continue
			}

			xCnt, ok := mem[x]

			if ok {
				val += (xCnt * mem[y]) % MOD
			}
		}
		mem[z] = val
		total = (total + val) % MOD
	}
	return total
}

File: python/medium/823_binary_trees_with_factors.py
class Solution:
    def numFactoredBinaryTrees(self, arr: List[int]) -> int:
        MOD = 1_000_000_007

        arr.sort()
        mem = {}
        output = 0
        for i, z in enumerate(arr):
            val = 0
            for x in arr[:i]:
                y, r = divmod(z, x)
                if not r and y in mem:
                    val += mem[x] * mem[y] % MOD

            val = (val + 1) % MOD
            output = (output + val) % MOD
            mem[z] = val
        return output
